File: ios/Anemone SDK/model/config/Property.swift
//
//  Property.swift
//  Anemone SDK
//
//  Created by Manuel Gonzalez Villegas on 9/12/16.
//  Copyright © 2016 Syltek Solutions S.L. All rights reserved.
//

import Foundation

public typealias PropertyId = Id

public struct Property {
    public static let durationPropertyId = PropertyId("duration")
    public static let sizePropertyId = PropertyId("resource_size")
    public static let typePropertyId = PropertyId("resource_type")
    public static let featurePropertyId = PropertyId("resource_feature")

    public let id: PropertyId
    public let name: String
    public let options: [PropertyOption]

    public func optionWithId(_ id: PropertyOptionId) -> PropertyOption? {
        options.filter { $0.id == id }.first
    }
}

extension Property: JSONMappable {
    public init(json: JSONObject) throws {
        self.id = try PropertyId(json.getAny("property_id"))
        self.name = try json.getString("name")
        self.options = try json.getJSONArray("options").flatMap { (json: JSONObject) in
            try PropertyOption(json: json)
        }
    }
}

extension Property: Equatable { }
public func == (lhs: Property, rhs: Property) -> Bool {
    lhs.id == rhs.id
}

File: android/app/src/main/java/com/playtomic/test/anemone_sdk/model/config/Property.kt
package com.anemonesdk.model.config

import com.anemonesdk.general.json.JSONMappable
import com.anemonesdk.general.json.JSONObject
import com.anemonesdk.model.generic.Id
import org.json.JSONException

/**
 * Created by agarcia on 23/12/2016.
 */
typealias PropertyId = Id

class Property : JSONMappable {
    companion object {
        val durationPropertyId = PropertyId("duration")
        val sizePropertyId = PropertyId("resource_size")
        val typePropertyId = PropertyId("resource_type")
        val featurePropertyId = PropertyId("resource_feature")
    }

    val id: PropertyId
    val name: String
    val options: List<PropertyOption>

    constructor(id: PropertyId, name: String, options: List<PropertyOption>) {
        this.id = id
        this.name = name
        this.options = options
    }

    @Throws(JSONException::class)
    constructor(json: JSONObject) : super(json) {
        id = PropertyId(json.getAny("property_id"))
        name = json.getString("name")
        options = json.getJSONArray("options").flatMap { f -> PropertyOption(f) }
    }

    fun optionWithId(id: PropertyOptionId): PropertyOption? {
        return options.filter { it.id == id }.firstOrNull()
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        return (other as? Property)?.id == id
    }

    override fun hashCode(): Int = id.hashCode()
}
